### ISSUE TITLE
Salt modules don't use minion pillar data for configuration when called from the master

### ISSUE THREAD

**Issue by kveroneau** [saltstack/salt#46410]
### Description of Issue/Question
Salt modules don't use minion pillar data for configuration when called from the master.

### Setup
Install either redis-server or mysql-server or both on a minion.  Set minion pillar for either `mysql.pass` or `redis.host` or both.  Run either the redis or mysql module from the master or on the minion.  Take notice that when running `salt-call` on the minion, it connects to the local minion redis or mysql server as expected.  Now go to the master and run the modules targeting the minion and realize that it is not using the pillars as it should for these modules, but rather the default values.

### Steps to Reproduce Issue
On minion run: `salt-call redis.keys` or `salt-call mysql.status` with correct pillar data set.

On master run: `salt 'minion' redis.keys` or `salt 'minion' mysql.status` with correct pillar data set.

### Versions Report
`
Salt Version:
           Salt: 2017.7.3
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.2
      docker-py: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.9 (default, Sep 17 2016, 20:26:04)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.0 
         locale: UTF-8
        machine: armv7l
        release: 4.9.35-v7+
         system: Linux
        version: debian 8.0 
`

**Comment by kveroneau** [comment]
Output from /var/log/salt/minion which aided me in figuring out why calling the modules from the master weren't working, but using `salt-call` on the minion did work:
``2018-03-06 20:47:24,327 [salt.log.setup   ][ERROR   ][7833] An un-handled exception was caught by salt's global exception handler:
ConnectionError: Error 110 connecting salt:6379. Connection timed out.
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 379, in salt_call
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/call.py", line 58, in run
    caller.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 134, in run
    ret = self.call()
  File "/usr/lib/python2.7/dist-packages/salt/cli/caller.py", line 197, in call
    ret['return'] = func(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/salt/modules/redismod.py", line 301, in keys
    return server.keys(pattern)
  File "/usr/lib/pymodules/python2.7/redis/client.py", line 455, in keys
    return self.execute_command('KEYS', pattern)
  File "/usr/lib/pymodules/python2.7/redis/client.py", line 283, in execute_command
    connection.send_command(*args)
  File "/usr/lib/pymodules/python2.7/redis/connection.py", line 260, in send_command
    self.send_packed_command(self.pack_command(*args))
  File "/usr/lib/pymodules/python2.7/redis/connection.py", line 243, in send_packed_command
    self.connect()
  File "/usr/lib/pymodules/python2.7/redis/connection.py", line 191, in connect
    raise ConnectionError(self._error_message(e))
ConnectionError: Error 110 connecting salt:6379. Connection timed out.
2018-03-06 20:50:07,690 [salt.loader.salt.int.module.mysql                    ][ERROR   ][9847] MySQL Error 1045: Access denied for user 'root'@'localhost' (using password: NO)
``

**Comment by kveroneau** [comment]
Upgraded salt to 2017.7.4 and issue persists.

**Comment by kveroneau** [comment]
Okay, after running `salt 'minion' saltutil.refresh_pillar` calls using `salt 'minion' mysql.status` worked.  Not sure if this is proper behavior or not, as I did run `pillar.items` on the minion, which I believe should refresh the pillars on the minion side.  I'll leave this issue opened for now until a proper explanation is provided on this behavior.

**Comment by garethgreenaway** [comment]
@kveroneau Thanks for the report.  `pillar.items` will refresh the pillar items but calls to other modules will not, with the exception of state.sls, state.highstate and state.apply.  Calling salt-call will also update pillar values.

To be clear, you're setting values in pillar for redis and mysql, then calling the `redis.keys` and `mysql.status` functions on the Master targeting your minion.  And not seeing the expected pillar values?

**Comment by kveroneau** [comment]
@garethgreenaway Yes, that is correct.  I did run pillar.items from the minion to confirm the redis and mysql values were available to the minion through pillar, however it wasn't until I ran saltutil.refresh_pillar that the master was able to use that pillar data, so it seems that pillar.items has a similar issue.  I haven't tried using mysql or redis yet in a state file to confirm if the same occurs there, but through the execution modules it definately appears.  I feel most people who use these execution modules will run them from the master targetting the minion, and will be equally confused on why they aren't working as expected.  I am also using the default pillar through /srv/pillars, no ext_pillar is being used for this.

**Comment by kveroneau** [comment]
I should also mention that running pillar.items on the minion doesn't refresh the pillar on the master, neither does salt-call on the minion.  For example, if you run `salt 'minion' redis.keys` on the master, then run `salt-call redis.keys` on the minion, then return back to the master and run `salt 'minion' redis.keys` it still doesn't work on the master until I ran saltutil.refresh_pillars.  This is where I got really confused and needed to resort to the minion log file to see what was going on.  So running `salt-call` from the minion doesn't seem to refresh the pillar data the master uses for some reason, although I thought the pillar data was parsed on the minion side all the time.
